### PR TITLE
Fix: Handle null albums, people, and excluded albums

### PR DESCRIPTION
**Pull request by jrh3k5**
If I provided an empty configuration for albums, people, or excluded albums, the deserialized settings appear to leave the respective `.Albums`, `.People,` and `.ExcludedAlbums` properties as `null`. This causes errors like the one below when ImmichFrame tries to iterate over them:

```
25-12-28 20:20:26 fail: Microsoft.AspNetCore.Server.Kestrel[13] Connection id "0HNI68OJKQ412", Request id "0HNI68OJKQ412:00000002": An unhandled exception was thrown by the application. System.ArgumentNullException: Value cannot be null. (Parameter 'source')    at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)    at System.Linq.Enumerable.TryGetNonEnumeratedCount[TSource](IEnumerable`1 source, Int32& count)    at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source)    at ImmichFrame.Core.Logic.PooledImmichFrameLogic.BuildPool(IAccountSettings accountSettings) in /source/ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs:line 54    at ImmichFrame.Core.Logic.PooledImmichFrameLogic..ctor(IAccountSettings accountSettings, IGeneralSettings generalSettings, IHttpClientFactory httpClientFactory) in /source/ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs:line 28    at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)    at System.Reflection.ConstructorInvoker.InvokeDirectByRefWithFewArgs(Span`1 copyOfArgs)    at System.Reflection.ConstructorInvoker.InvokeDirectByRef(Object arg1, Object arg2, Object arg3, Object arg4)    at System.Reflection.ConstructorInvoker.InvokeImpl(Object arg1, Object arg2, Object arg3, Object arg4)    at System.Reflection.ConstructorInvoker.Invoke(Span`1 arguments)    at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)    at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)    at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance[T](IServiceProvider provider, Object[] parameters)    at Program.<>c__DisplayClass0_1.<<Main>$>b__8(IAccountSettings account) in /source/ImmichFrame.WebApi/Program.cs:line 67    at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)    at System.Collections.Frozen.FrozenDictionary.ToFrozenDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)    at ImmichFrame.Core.Logic.MultiImmichFrameLogicDelegate..ctor(IServerSettings serverSettings, Func`2 logicFactory, ILogger`1 logger, IAccountSelectionStrategy accountSelectionStrategy) in /source/ImmichFrame.Core/Logic/MultiImmichFrameLogicDelegate.cs:line 23    at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)    at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)    at System.Reflection.MethodBaseInvoker.InvokeWithFewArgs(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)    at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)    at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)    at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)    at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)    at lambda_method18(Closure, IServiceProvider, Object[])    at Microsoft.AspNetCore.Mvc.Controllers.ControllerFactoryProvider.<>c__DisplayClass6_0.<CreateControllerFactory>g__CreateController|0(ControllerContext controllerContext)    at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)    at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync() --- End of stack trace from previous location ---    at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)    at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)    at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)    at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)    at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)    at CustomAuthenticationMiddleware.InvokeAsync(HttpContext context) in /source/ImmichFrame.WebApi/Helpers/CustomAuthenticationMiddleware.cs:line 23    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```

This adds `null` awareness around uses of these enumerables to protect against these errors.

## Summary by CodeRabbit

* **Bug Fixes**
  * Enhanced null-safety handling for account settings collections to prevent potential runtime errors.

* **Tests**
  * Added test coverage for edge cases with null collection configurations.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs [hunk 1]
```diff
@@ -35,7 +35,13 @@ private static TimeSpan RefreshInterval(int hours)
 
     private IAssetPool BuildPool(IAccountSettings accountSettings)
     {
-        if (!accountSettings.ShowFavorites && !accountSettings.ShowMemories && !accountSettings.Albums.Any() && !accountSettings.People.Any())
+        var albums = accountSettings.Albums;
+        var hasAlbums = albums != null && albums.Any();
+
+        var people = accountSettings.People;
+        var hasPeople = people != null && people.Any();
```
Comment (JW-CH) [inline ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs]:
```suggestion
        var hasPeople = accountSettings.People?.Any() ?? false;
```

File: ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs [hunk 2]
```diff
@@ -35,7 +35,13 @@ private static TimeSpan RefreshInterval(int hours)
 
     private IAssetPool BuildPool(IAccountSettings accountSettings)
     {
-        if (!accountSettings.ShowFavorites && !accountSettings.ShowMemories && !accountSettings.Albums.Any() && !accountSettings.People.Any())
+        var albums = accountSettings.Albums;
+        var hasAlbums = albums != null && albums.Any();
```
Comment (JW-CH) [inline ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs]: I'd prefer something like this:


```suggestion
        var hasAlbums = accountSettings.Albums?.Any() ?? false;
```

File: ImmichFrame.Core/Logic/Pool/AllAssetsPool.cs [hunk 3]
```diff
@@ -50,7 +50,8 @@ public async Task<IEnumerable<AssetResponseDto>> GetAssets(int requested, Cancel
 
         var assets = await immichApi.SearchRandomAsync(searchDto, ct);
 
-        if (accountSettings.ExcludedAlbums.Any())
+        var excludedAlbums = accountSettings.ExcludedAlbums;
+        if (excludedAlbums != null && excludedAlbums.Any())
```
Comment (JW-CH) [inline ImmichFrame.Core/Logic/Pool/AllAssetsPool.cs]:
```suggestion
        if (accountSettings.ExcludedAlbums?.Any() ?? false)
```

File: ImmichFrame.Core/Logic/Pool/AllAssetsPool.cs [hunk 4]
```diff
@@ -50,7 +50,8 @@ public async Task<IEnumerable<AssetResponseDto>> GetAssets(int requested, Cancel
 
         var assets = await immichApi.SearchRandomAsync(searchDto, ct);
 
-        if (accountSettings.ExcludedAlbums.Any())
+        var excludedAlbums = accountSettings.ExcludedAlbums;
```
Comment (JW-CH) [inline ImmichFrame.Core/Logic/Pool/AllAssetsPool.cs]: Not needed anymore.

File: ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs [hunk 5]
```diff
@@ -35,7 +35,13 @@ private static TimeSpan RefreshInterval(int hours)
 
     private IAssetPool BuildPool(IAccountSettings accountSettings)
     {
-        if (!accountSettings.ShowFavorites && !accountSettings.ShowMemories && !accountSettings.Albums.Any() && !accountSettings.People.Any())
+        var albums = accountSettings.Albums;
```
Comment (JW-CH) [inline ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs]: Not needed anymore.

File: ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs [hunk 6]
```diff
@@ -35,7 +35,13 @@ private static TimeSpan RefreshInterval(int hours)
...
+        var hasAlbums = accountSettings.Albums?.Any() ?? false;
+
+        var people = accountSettings.People;
```
Comment (JW-CH) [inline ImmichFrame.Core/Logic/PooledImmichFrameLogic.cs]: Not needed anymore.

**Comment by jrh3k5** [comment]
@JW-CH - sorry, I should have previewed these changes in the IDE before committing them - the 'unused variable' warnings would have been more visible.

I've also replaced the calls to `.Any()` with comparisons against the `.Count` field - the in-IDE linter said that was more performant.

**Comment by JW-CH** [comment]
> @JW-CH - sorry, I should have previewed these changes in the IDE before committing them - the 'unused variable' warnings would have been more visible.
> 
> I've also replaced the calls to `.Any()` with comparisons against the `.Count` field - the in-IDE linter said that was more performant.

Sorry for the back and forth. I just did some reading (on a quite outdated article, lol) https://www.tabsoverspaces.com/233649-comparing-speed-of-count-0-and-any

But only the numbers should be outdated, not the point this guy is making: 
> ... I write readable code first, often that also means performant enough, and optimize later when needed and backed by numbers.

In this case, I quite agree with that and would ignore the lesser performance and value readability higher; it also fits the style of the rest better. What are your thoughts? (By the way, LINQ got better over time and more optimized. I don't think it will be much of a difference, plus the love LINQ got performance-wise in .NET 10 is incredible (I know, Frame is on .NET 8 as of now))

**Comment by jrh3k5** [comment]
> What are your thoughts?

My thoughts are it's been a very hot minute since I've written .NET professionally. 😄 I'm fine with `.Any() ?? false` if that's more readable and consistent with the codebase, and have updated the PR accordingly.

**Comment by jrh3k5** [comment]
Hey, @JW-CH - apologies if I'm being pushy, but I also know that it's easy for PR notifications to get lost in the shuffle. Just want to nudge that this PR is ready for review.

**Comment by 3rob3** [comment]
I didn't review the tests, but the logic looks good to me. 👍

**Comment by JW-CH** [comment]
Sorry, I lack behind with reviews. Hopefully, I'll find this weekend :)

File: ImmichFrame.Core.Tests/Logic/PooledImmichFrameLogicTests.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    [Test]
+    public void BuildPool_WithNullPeople_DoesNotAddPersonAssetsPool()
+    {
+        // Arrange
+        _mockAccountSettings.SetupGet(s => s.People).Returns((List<Guid>)null);
+
+        // Act
+        var logic = new PooledImmichFrameLogic(_mockAccountSettings.Object, _mockGeneralSettings.Object, _mockHttpClientFactory.Object);
+
+        // Assert - The pool should be AllAssetsPool, not MultiAssetPool
```
Comment (JW-CH) [inline ImmichFrame.Core.Tests/Logic/PooledImmichFrameLogicTests.cs]: these comments are quite off? can you check this file again?

Comment (JW-CH) [reply ImmichFrame.Core.Tests/Logic/PooledImmichFrameLogicTests.cs]: Also some of these checks do not really do anything? Checking if a constructor added something successfully does not make much sense to me.

File: ImmichFrame.Core.Tests/Logic/PooledImmichFrameLogicTests.cs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+    /// <summary>
+    /// Ensures that mixed settings (some enabled, some disabled) with null People
+    /// creates a proper MultiAssetPool excluding PersonAssetsPool
+    /// </summary>
+    [Test]
+    public void BuildPool_WithMixedSettingsAndNullPeople_DoesNotIncludePersonAssetsPool()
+    {
+        // Arrange - enable some settings but keep People null
+        _mockAccountSettings.SetupGet(s => s.ShowFavorites).Returns(true);
```
Comment (JW-CH) [inline ImmichFrame.Core.Tests/Logic/PooledImmichFrameLogicTests.cs]: why are these set up?

File: ImmichFrame.Core.Tests/Logic/PooledImmichFrameLogicTests.cs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        _mockAccountSettings.SetupGet(s => s.ShowMemories).Returns(false);
+        _mockAccountSettings.SetupGet(s => s.Albums).Returns(new List<Guid>());
+        _mockAccountSettings.SetupGet(s => s.People).Returns((List<Guid>)null);
+
+        // Act
+        var logic = new PooledImmichFrameLogic(_mockAccountSettings.Object, _mockGeneralSettings.Object, _mockHttpClientFactory.Object);
+
+        // Assert
+        Assert.That(logic.AccountSettings.People, Is.Null);
```
Comment (JW-CH) [inline ImmichFrame.Core.Tests/Logic/PooledImmichFrameLogicTests.cs]: wouldn't this throw if you access `logic.AccountSettings.People` if `logic.AccountSettings` is null? Asking because of the next check a line below.

**Comment by jrh3k5** [comment]
@JW-CH - apologies, I did export the work on the tests to Copilot. I should have looked more closely at the tests before submitting the PR.

With that in mind, I'm trying to do some TDD to verify that the tests actually catch errors, and I'm struggling with Moq here. In this case, I've removed the calls to `.SetupGet()` in the `Setup()` method (so that the default setup of returning non-null lists wouldn't interfere with the test and commented out the null check, expecting it to throw a null pointer exception, and it doesn't.

image

We know that it _can_ be `null` at runtime because of the stacktrace attached in the PR description, so I suspect that Moq is trying to "help" here and return a non-null default value.

If that's the case, I could possibly try to get around it by maybe creating a concrete instance of the settings and, within the scope of the test, create a new instance of `TestableAlbumAssetsPool`? But I generally hate doing that because it makes changes to constructors a multiplicative effort for tests (and I've seen it more than once lead to poorly-maintained tests).

I apologize, again, for wasting your time with my AI-generated unit tests.

**Comment by JW-CH** [comment]
> I apologize, again, for wasting your time with my AI-generated unit tests.

No need to apologize, I like reviewing stuff, but this takes time :) I'm glad that you like the project and want to make changes yourself. If you can't get the tests to work, please just remove them. I'd prefer having no tests over AI-gibberish tests. I can also add some myself.

I have not the greatest knowledge myself in testing software, that's why I do not have a strong opinion about this.

**Comment by jrh3k5** [comment]
@JW-CH - figured out the unlock: `.Returns()` can accept a lambda, which allows me to bypass whatever Moq logic was preventing me from returning `null`. I've got all the tests working and also removed the AI slop unit tests here:

2207031b456121cac44de9727396ac3742074f41

**Review by JW-CH — APPROVED** [review]
I have pushed on your PR, I hope you don't mind. If you want to return `null` in the setup, you are able to do it like so:

```cs
_mockAccountSettings.SetupGet(s => s.ExcludedAlbums).Returns((List<Guid>)null);
```

I assume that was what you tried to do with your added property :) Thank you for your PR!

**Comment by jrh3k5** [comment]
@JW-CH:

> I have pushed on your PR, I hope you don't mind. If you want to return null in the setup, you are able to do it like so:

I found that actually didn't work. The reason I had to make the change to a property-driven provisioning of the albums reference is that, if I _removed_ the null checks I added _and_ used the `.SetupGet()` as your change does, it did not actually reproduce the error - Moq seems to silently ignore such setups. As a result, `.SetupGet(...).Returns((List<Guid>) null))` _actually_ ends up returning a non-null, empty `List<Guid>` instance.

It's not a hill I'll die on, but I do want to stress that the tests, as they exist in `main`, now, do not actually test for when `ExcludedAlbums`, `Albums`, and `People` are `null` - the tests are all false positives because, despite what the tests are setting up, Moq is actually setting up a return of non-null, empty lists.

**Comment by JW-CH** [comment]
> @JW-CH:
> 
> > I have pushed on your PR, I hope you don't mind. If you want to return null in the setup, you are able to do it like so:
> 
> I found that actually didn't work. The reason I had to make the change to a property-driven provisioning of the albums reference is that, if I _removed_ the null checks I added _and_ used the `.SetupGet()` as your change does, it did not actually reproduce the error - Moq seems to silently ignore such setups. As a result, `.SetupGet(...).Returns((List<Guid>) null))` _actually_ ends up returning a non-null, empty `List<Guid>` instance.
> 
> It's not a hill I'll die on, but I do want to stress that the tests, as they exist in `main`, now, do not actually test for when `ExcludedAlbums`, `Albums`, and `People` are `null` - the tests are all false positives because, despite what the tests are setting up, Moq is actually setting up a return of non-null, empty lists.

Good to know! Thank you! I'll have rework the tests sometimes in the future again. Will keep this in mind. For now, I'll think we are ok.
